quinn-proto/src/connection/streams/recv.rs

```diff
@@ -257,6 +247,89 @@ impl Recv {
     }
 }
 
+pub struct Chunks<'a> {
+    id: StreamId,
+    ordered: bool,
+    streams: &'a mut Streams,
+    pending: &'a mut Retransmits,
+    recv: Option<Recv>,
+}
+
+impl<'a> Chunks<'a> {
+    pub(super) fn new(
+        id: StreamId,
+        ordered: bool,
+        streams: &'a mut Streams,
+        pending: &'a mut Retransmits,
+    ) -> Result<Self, ReadableError> {
+        let entry = match streams.recv.entry(id) {
+            Entry::Occupied(entry) => entry,
+            Entry::Vacant(_) => return Err(ReadableError::UnknownStream),
+        };
+
+        let mut recv = match entry.get().stopped {
+            true => return Err(ReadableError::UnknownStream),
+            false => entry.remove(),
+        };
+
+        recv.assembler.ensure_ordering(ordered)?;
+        Ok(Self {
+            id,
+            ordered,
+            streams,
+            pending,
+            recv: Some(recv),
+        })
+    }
+
+    pub(crate) fn next(&mut self, max_length: usize) -> Result<Option<Chunk>, ReadError> {
+        let mut rs = self.recv.take().unwrap();
+        if let Some(chunk) = rs.assembler.read(max_length, self.ordered) {
+            let (_, max_stream_data) = rs.max_stream_data(self.streams.stream_receive_window);
+            let max_data = self.streams.add_read_credits(chunk.bytes.len() as u64);
+            self.pending
+                .post_read(self.id, max_data, max_stream_data, false);
+            self.streams.recv.insert(self.id, rs);
```

Comment:
Within the scope of this specific commit, this doesn't make sense; we should put `rs` back into `self.recv` so futures calls to `next` will succeed.